File: Linked_Lists/Reverse_Linked_List.py
# Given the head of a singly linked list, reverse the list, and return the reversed list.

# Example 1:
# Input: head = [1,2,3,4,5]
# Output: [5,4,3,2,1]

# Solution: Iterate through linked list, set prev to None curr to head keep track of next ptr of curr, reverse
# curr and move ptrs forward

def reverseList(self, head: ListNode) -> ListNode:
    # Intialize variables, prev is going to be the new head
    prev, curr = None, head

    # Iterate through linked list
    while curr:
        # Temp variable to store next ptr since we lose reference to it
        nextTemp = curr.next
        # Reverse
        curr.next = prev
        # Move ptrs forward
        prev = curr
        curr = nextTemp

    return prev

# Time Complexity: O(N)
# Space Complexity: O(1)

File: Linked_Lists/Reverse_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* next_temp = nullptr;
        ListNode* curr = head;
        while(curr) {
            next_temp = curr -> next;
            curr -> next = prev;
            prev = curr;
            curr = next_temp;
        }
        return prev;
    }
};